pennylane/utils.py

```diff
@@ -393,3 +393,54 @@ def expand_vector(vector, original_wires, expanded_wires):
     )
 
     return qml.math.reshape(expanded_tensor, 2**M)
+
+
+def prod_sort(op_list, wire_map: dict = None):
+    """Insertion sort algorithm that sorts a list of product factors by their wire indices, taking
+    into account the operator commutivity.
+
+    Args:
+        op_list (List[.Operator]): list of operators to be sorted
+        wire_map (dict): Dictionary containing the wire values as keys and its indexes as values.
+            Defaults to None.
+
+    Returns:
+        List[.Operator]: sorted list of operators
+    """
+
+    def swappable_ops(op1, op2, wire_map: dict) -> bool:
+        """Boolean expression that indicates if op1 and op2 are commutative and should be swapped when
+        sorting them by wire values.
+
+        Args:
+            op1 (.Operator): First operator.
+            op2 (.Operator): Second operator.
+            wire_map (dict): Dictionary containing the wire values as keys and its indexes as values.
+                Defaults to None.
+
+        Returns:
+            bool: True if operators should be swapped, False otherwise.
+        """
+        wires1 = op1.wires
+        wires2 = op2.wires
+        if wire_map is not None:
+            wires1 = wires1.map(wire_map)
+            wires2 = wires2.map(wire_map)
+        if np.intersect1d(wires1, wires2).size != 0:
+            return False
```

Comment:
Are we purposefully ignoring the scenario where two operators commute and have common wires?

Reply:
```
>>> ops = [qml.CZ((0,1)), qml.CZ((-1, 0))]
>>> qml.is_commuting(*ops)
True
>>> qml.utils.prod_sort(ops)
[CZ(wires=[0, 1]), CZ(wires=[-1, 0])]
```

Reply:
Right, but my question is more about whether this is an intentional design choice. For example, with the following list of observables:
```python
[qml.PauliZ(1) @ qml.PauliZ(2) @ qml.PauliX(3), qml.PauliX(0) @ qml.PauliY(1) @ qml.PauliY(2)]
```
since the two elements commute, I would expect the sorted list to be:
```python
[qml.PauliX(0) @ qml.PauliY(1) @ qml.PauliY(2), qml.PauliZ(1) @ qml.PauliZ(2) @ qml.PauliX(3)]
```

Reply:
I wanted to make the sorting condition as simple as possible to avoid adding a lot of time complexity in the simplify pipeline. I realised that the `qml.is_commuting` function is uses lookup tables, thus it should be quite fast for non-rotating operators. However the check is not supported for operator arithmetic (Prod, Sum...).

Reply:
For now we could use this simple check and then later we could add support of op. arithmetic to `qml.is_commuting` and use that instead.